File: javascript/meetup/meetup.js
'use strict';
const days = ['Sunday', 'Monday', 'Tuesday',
  'Wednesday', 'Thursday', 'Friday', 'Saturday'];
const m = {'1st': 0, '2nd': 1, '3rd': 2, '4th': 3, '5th': 4};

module.exports = function(year, month, day, instance) {
  day = 1 + days.indexOf(day) - (new Date(year, month, 1)).getDay();
  let max = (new Date(year, month + 1, 0)).getDate();
  if (instance === 'teenth') {
    while (day < 10)
      day += 7;
  } else if (instance in m) {
    let x = (day - 1 < 0) ? 1 : 0;
    x += m[instance];
    day += (7 * x);
    if (day > max)
      throw new Error('Invalid day.');
  } else if (instance === 'last') {
    while (day + 7 <= max)
      day += 7;
  }
  return new Date(year, month, day);
};

File: python/meetup/meetup.py
from datetime import date
import calendar


class MeetupDayException(Exception):
    pass


def meetup_day(year, month, day_name, descriptor):
    weekdays = dict(zip(list(calendar.day_name), range(0, 7)))
    descriptors = {"1st": 0, "2nd": 1, "3rd": 2, "4th": 3, "5th": 4}
    day = 1 + ((weekdays.get(day_name) - date(year, month, 1).weekday()) % 7)
    last_day = calendar.monthrange(year, month)[1]
    if descriptor == "teenth":
        while day + 7 < 20:
            day += 7
    elif descriptor == "last":
        while day + 7 <= last_day:
            day += 7
    else:
        day += 7 * descriptors[descriptor]
    if day > last_day:
        raise MeetupDayException()
    return date(year, month, day)
